docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -0,0 +1,442 @@
+  across enclaves. To solve the problem, this document uses an idea of storing the
+  list of ECALL id per enclave, which we refer to as local ECALL tables.
+  The local table is declared as follows.
+
+  ```
+  typedef struct _oe_ecall_id_t
+  {
+     uint64_t id;
+  } oe_ecall_id_t;
+
+  typedef struct _oe_enclave
+  {
+     ...
+     /* Per-edl ecall id table. */
+     oe_ecall_id_t ecall_id_table[OE_ECALL_MAX];
+     uint32_t ecall_id_table_size;
+     ...
+  } oe_enclave_t;
+  ```
+
+  The local table allows the host to retrieve the ECALL id per enclave instead
+  of using the hard-coded id in the same ECALL across different enclaves.
+
+- Global Table
+
+  To guarantee the constant-time look-up, the host needs to use the same id to retrieve
+  the ECALL id from the local table. To do so, the host additionally maintains a global
+  table that keeps such ids to which we refer as global ids. The detailed design is as follows.
+
+  - The global table is defined internally in the `oehost` as follows. The table stores the name
+    of each ECALL and the index to each name is used as the global id.
+    ```
+    const char* _ecall_table[OE_ECALL_MAX];
+    uint32_t _ecall_table_size;
+    ```
+  - The global table is updated during every initialization of a local table. More specifically,
+    the list of ECALL names is added to the `_u.c` and is passed to the `oe_create_enclave`.
+    ```
+    typedef struct _oe_ecall_info_t
+    {
+        const char* name;
+    } oe_ecall_info_t;
+
+    /**** Trusted function names. ****/
+    const oe_ecall_info_t __foo_ecall_info_table[] =
+    {
+        { "ecall_common_1_ecall" },
+        { "ecall_common_2_ecall_1" },
+        { "ecall_common_2_ecall_2" },
+        { "ecall_foo_ecall" },
+        { NULL }
+    };
+    ```
+    ```
+    oe_result_t oe_create_foo_enclave(
+      const char* path,
+      oe_enclave_type_t type,
+      uint32_t flags,
+      const oe_enclave_setting_t* settings,
+      uint32_t setting_count,
+      oe_enclave_t** enclave)
+    {
+        return oe_create_enclave(
+                 path,
+                 type,
+                 flags,
+                 settings,
+                 setting_count,
+                 __foo_ocall_function_table,
+                 3,
+                 __foo_ecall_info_table,
+                 3,
+                 enclave);
+    }
+    ```
+  - During the initialization of the local table, the host uses the following logic
+    to get the global id by name.
+    ```
+    uint32_t i, index;
+    for (i = 0; i < ecall_info_table_size; i++)
+    {
+      if found name in the global_table
+        index = the index of the name
+      else /* not found */
+        add the name to the table
+        index = the index of the name
+
+      /* Assign the ECALL id to the local table based on the global id. */
+      enclave->ecall_id_table[index] = i;
+    }
+    ```
+  - Modify the internal `oe_call_enclave_function` function to pass a global id
```

Comment:
@dthaler  Please refer to here for the use of static variable to cache the global id after the first invocation of an ecall.